### DISCUSSION TITLE
Using mdx provider in a server component

### DISCUSSION THREAD

**Discussion by gmcamposano** [vercel/next.js#50897] [Help]
### Summary

1) I first tried to implement it on the page, just as the next docs example, but it fails as it has to be a client component which at the same time can't be done due to the fact that it has meta data: https://nextjs.org/docs/app/building-your-application/configuring/mdx#custom-elements
2) I tried a wrapper using a client component inside the wrapper, but elements are ignored. In my case i am styling an h1 and trying to pass next Image to all images.
3) I even tried importing components inside the MDX files and I get the message: to use ____ you must first import the component

So basically, how can i handle my custom components here:

```
'use client'
import ImageLoader from '@/helpers/image-loader'
import { MDXProvider } from '@mdx-js/react'

const components = {
  h1: ({ children }) => <h1 style={{ color: '#2c2' }}>{children}</h1>,
  img: ({ src, alt, ...props }) => {
    return <ImageLoader src={src} alt={alt} {...props} />
  },
}

export default async function Wrapper({ children }) {
  return <MDXProvider components={components}>{children}</MDXProvider>
}

```

page.js

```
import Wrapper from './wrapper'
import { getPostBySlug } from '@/helpers/blogPostsFetch'
import Wrapper 

const getPageContent = async (slug) => {
  const { content, meta } = await getPostBySlug(slug)
  return { content, meta }
}

export async function generateMetadata({ params }) {
  const { meta } = await getPageContent(params.slug)
  return {
    title: meta.title,
  }
}

export default async function BlogSlug({ params }) {
  const { content } = await getPageContent(params.slug)
  return (
<Wrapper >
    <section>
        {content}
    </section>
</Wrapper>
  )
}

```

next.config.mjs:
```
import nextMDX from '@next/mdx'

/** @type {import('next').NextConfig} */
const nextConfig = {
  pageExtensions: ['js', 'jsx', 'ts', 'tsx', 'md', 'mdx'],
  experimental: {
    mdxRs: true,
  },
  images: {
    domains: ['res.cloudinary.com', 'images.unsplash.com', 'unsplash.it'],
  },
}

const withMDX = nextMDX({
  options: {
    remarkPlugins: [],
    rehypePlugins: [],
    providerImportSource: '@mdx-js/react',
  },
})
// module.exports = withMDX(nextConfig)
export default withMDX(nextConfig)

```

**Comment by icyJoseph** [comment]
Couldn't the components live in another module which could use the 'use client' directive?

Also in your first block, `Wrapper` can't be async because it is a Client Component.

I didn't use the same package, but here you can see the next-remote-mdx approach:

- All components: https://github.com/icyJoseph/icyJoseph/blob/dev/src/components/Blog/mdx.tsx 
- CodeBlock which is a client component https://github.com/icyJoseph/icyJoseph/blob/dev/src/components/Blog/CodeBlock.tsx
- Using it at page (server component) level: https://github.com/icyJoseph/icyJoseph/blob/dev/src/app/blog/%5Bslug%5D/page.tsx

**Comment by gmcamposano** [comment]
Hello,
So I changed the wrapper to live inside the layout (using the 'use client' directive) and MDX Provider is simply not being read. 

One case of this use I want to do is all images have to use the Image component, list styles this way.. etc. That way I dont have to manually repeat the same thing over and over again in the MDX files. 
layout.js
```
'use client'
import { MDXProvider } from '@mdx-js/react'
import Image from 'next/image'

const components = {
  h1: ({ children }) => <h1 style={{ color: '#2c2' }}>{children}</h1>,
  img: (props) => <Image {...props} />,
}

export default function Layout({ children }) {
  return <MDXProvider components={components}>{children}</MDXProvider>
}
```
I have also included a mdx-components.js
```
export function useMDXComponents(components) {
  return {
    h1: ({ children }) => (
      <h1 style={{ fontSize: '500px', color: 'green' }}>{children}</h1>
    ),
    ...components,
  }
}

```
But all it keeps ignoring the configurations I do and just render the content without applying the style.

**Reply by icyJoseph** [reply]
Got a repository to take a look at?

**Reply by gmcamposano** [reply]
I made a sample doc. Here you go
https://codesandbox.io/p/sandbox/late-fog-pugr5o?file=%2Fposts%2Fhello.mdx%3A11%2C57

**Reply by icyJoseph** [reply]
And you want something like this, through the MDX Provider, right?

![Screenshot 2023-06-08 at 14 45 11](https://github.com/vercel/next.js/assets/21013447/3b5890bd-59b9-49c4-a7d2-32e33e7c754e)

**Reply by icyJoseph** [reply]
Alright, I think I've collected my thoughts here.

Next @next/mdx plugin, seems to be most useful to load mdx like this:

```tsx
"use client";
import Hello from "../hello.mdx";

export default function Page() {
  return <Hello />;
}
```

Then I see the styles from components.

However this requires this file structure:

```console
app
├── Provider.tsx
├── components.tsx
├── favicon.ico
├── globals.css
├── layout.tsx
├── lib
│   └── fetch.tsx
├── page.tsx
└── post
    ├── hello
    │   └── page.tsx
    ├── hello.mdx
    └── second-post.mdx
```

Where `app/post/hello/page.tsx` is re-exporting the `hello.mdx` file, as a React component.

For completeness sake, heres the `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  images: {
    domains: ["unsplash.it"],
  },
  experimental: {
    mdxRs: true,
  },
};

const withMDX = require("@next/mdx")({
  options: { providerImportSource: "@mdx-js/react" },
});
module.exports = withMDX(nextConfig, {
  pageExtensions: ["md", "mdx", "tsx", "ts", "jsx", "js"],
});
```

The layout:

```tsx
import { Provider } from "./Provider";
import "./globals.css";
import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <Provider>{children}</Provider>
      </body>
    </html>
  );
}
```

The Provider:

```js
"use client";

import { MDXProvider } from "@mdx-js/react";
import { components } from "./components";

export function Provider({ children }: { children: React.ReactNode }) {
  return <MDXProvider components={components}>{children}</MDXProvider>;
}
```

And the components:

```tsx
import type { ComponentPropsWithoutRef } from "react";

export const components = {
  h1: (props: ComponentPropsWithoutRef<"h1">) => (
    <h1 style={{ color: "red" }} {...props} />
  ),
};
```

## Mixing tech stacks

My question here is, why though? `@mdx-js/react` has its own compiler, which `@next/mdx` implements under the hood. Meanwhile in `lib/fetch.tsx` you compile your MDX with `next-mdx-remote/rsc`, which is a different tech stack.

When you compile with that stack you can just do:

```ts
import fs from "fs";
import path from "path";
import { compileMDX } from "next-mdx-remote/rsc";
import { components } from "../components";

const FOLDER_PATH = path.join(process.cwd(), "posts");

export const getPostBySlug = async (slug) => {
  const fileSlug = slug.replace(/\.mdx$/, "");
  const fileContent = fs.readFileSync(`${FOLDER_PATH}/${fileSlug}.mdx`, {
    encoding: "utf-8",
  });
  const { frontmatter, content } = await compileMDX({
    source: fileContent,
    options: {
      parseFrontmatter: true,
    },
    components,
  });

  return {
    meta: { ...frontmatter, slug: fileSlug },
    content,
  };
};

// omitted getAllPosts
```

And then render with:

```tsx
import { getPostBySlug } from "../../lib/fetch";

const getPageContent = async (slug) => {
  const { content, meta } = await getPostBySlug(slug);
  return { content, meta };
};

export default async function BlogSlug({ params }) {
  const { content } = await getPageContent(params.slug);
  return (
    <section>
      <div>{content}</div>
    </section>
  );
}
```

No providers, no special `next.config.js`, nada, and it just works.

**Reply by gmcamposano** [reply]
Thanks! I will update my answer once I try it.
Why I did it like that? I followed a tutorial :D If there was another way i would also have done it that way.

**Reply by icyJoseph** [reply]
Yeah, I think the tutorial doesn't really make it clear where one tech stack or, rather, technique ends and the other starts. Let me know how it goes! There's a Lee Rob blog where he explains it way better. I'll try to find it.

- https://nextjs.org/blog/markdown

**Reply by gmcamposano** [reply]
Here you can have a look for context sake: https://youtu.be/n2CV6f0vFr4

**Reply by gmcamposano** [reply]
Regarding this code. It worked like a charm. But I am now more inclined in following your advised way of writing this. I will read the docs you sent of leerob and try to recreate it for you to evaluate. Thanks a lot and keep you posted.
```
import fs from "fs";
import path from "path";
import { compileMDX } from "next-mdx-remote/rsc";
import { components } from "../components";

const FOLDER_PATH = path.join(process.cwd(), "posts");

export const getPostBySlug = async (slug) => {
  const fileSlug = slug.replace(/\.mdx$/, "");
  const fileContent = fs.readFileSync(`${FOLDER_PATH}/${fileSlug}.mdx`, {
    encoding: "utf-8",
  });
  const { frontmatter, content } = await compileMDX({
    source: fileContent,
    options: {
      parseFrontmatter: true,
    },
    components,
  });

  return {
    meta: { ...frontmatter, slug: fileSlug },
    content,
  };
};

// omitted getAllPosts
```

**Reply by karlhorky** [reply]
> ## Mixing tech stacks
> 
> My question here is, why though? `@mdx-js/react` has its own compiler, which `@next/mdx` implements under the hood.

First of all, to talk about the downsides of `next-mdx-remote/rsc` or `mdx-bundler`:

- [`next-mdx-remote/rsc` doesn't allow for importing MDX files within MDX files](https://github.com/hashicorp/next-mdx-remote/issues/398#issuecomment-1999743798)
- [`mdx-bundler` requires complicated manual configuration to share components to imported `.mdx` files](https://github.com/kentcdodds/mdx-bundler/pull/227) (not to mention pulling in a bunch of extra bundling dependencies and running the bundling outside of the existing webpack / Turbopack bundler config)

Secondly, I think there are probably use cases for providing divergent MDX components in different areas (`mdx-components.tsx` from `@next/mdx` for all MDX files, and then `MDXProvider` from `@mdx-js/react` for providing the divergent components specific to a certain area).

Eg. we want this to implement different provided `components` in different areas of a large app, without having to do [`components={props.components}` prop drilling as suggested by the docs](https://nextjs.org/docs/app/building-your-application/configuring/mdx#using-custom-styles-and-components):

- https://github.com/vercel/next.js/issues/69613

---

## New `@next/mdx` API?

We would also be open to having an extra `MDXProvider`-style export from `@next/mdx` to be able to provide extra components at arbitrary locations and depths in the tree, in case that would be an option.

I could imagine that this may be an extension to the MDX components RSC support added by @timneutkens in PR #44651

**Comment by joshuaavalon** [comment]
I don't know if you still are looking for answer. But I found that MDX cannot be used in server component is caused by `@mdx-js/react` using `createContext`.

`@next/mdx` includes `@mdx-js/react` by default.

https://github.com/vercel/next.js/blob/87b66f64e53b9e5867cc88a7d648f4ab0af816ac/packages/next-mdx/index.js#L24-L28

To work around it, you can set `providerImportSource` to `private-next-root-dir/src/mdx-components` or `private-next-root-dir/mdx-components` to prevent loading `@mdx-js/react`.

Create `mdx-components.tsx` in the root of your application (the parent folder of app/ or src/).

Use the example from https://nextjs.org/docs/app/building-your-application/configuring/mdx. 

You **don't** need `MDXProvider` to pass in custom components. You can pass by `mdx-components.tsx` for global components and `<MdxTag components={components} />` for MDX-specific component.

```ts
import type { MDXComponents } from 'mdx/types'
 
// This file allows you to provide custom React components
// to be used in MDX files. You can import and use any
// React component you want, including components from
// other libraries.
 
// This file is required to use MDX in `app` directory.
export function useMDXComponents(components: MDXComponents): MDXComponents {
  return {
    // Allows customizing built-in components, e.g. to add styling.
    // h1: ({ children }) => <h1 style={{ fontSize: "100px" }}>{children}</h1>,
    ...components,
  }
}
```

Then you should able to import `.mdx` in server component.

**Reply by lsagetlethias** [reply]
What if you want an in-between kind of configuration?
Where you can have custom components only for one folder (`(routesGroup)` or sub routes) either by having a local `mdx-components.tsx` or by using a provider in the localised `layout.tsx`

**Reply by LordZardeck** [reply]
This is my current issue. My blog post articles will use different styles for headers, links, etc, and i'd like my layout within `/blog` to configure these overrides.

**Reply by gao-sun** [reply]
@gmcamposano you saved my day. thanks.

**Comment by Cliftonz** [comment]
Also having this issue

**Comment by landed1** [comment]
Not an answer but I don't see why this should be using a client for my purpose at least. I am trying to just create a static website and have previously done this with markdown but fail because of mdx needing to be a client component..
